dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/file/FileDownloadAPIHandler.java

```diff
@@ -264,158 +237,78 @@ private boolean isAccessAuthorized(DatasetVersion dsv, String apiToken) {
-                            .has(Permission.ViewUnpublishedDataset)) {
-                        logger.log(Level.FINE, "API token-based auth: User {0} has rights to access the (unpublished) datafile.", user
-                                .getIdentifier());
-                        return true;
-                    } else {
-                        logger.log(Level.FINE, "API token-based auth: User {0} is not authorized to access the (unpublished) datafile.", user
-                                .getIdentifier());
-                    }
-                }
-            } else {
-                logger.log(Level.FINE, "API token-based auth: User {0} is not authorized to access the datafile.", user.getIdentifier());
-            }
+        DataFile file = fileMetadata.getDataFile();
+        User apiTokenUser = getApiTokenUserWithGuestFallbackOnInvalidToken();
+        User sessionUser = getSessionUserWithGuestFallback();
 
-            boolean isWorldMapTokenPresentAndAuthorized = dsv.getFileMetadatas()
-                                                             .stream()
-                                                             .map(FileMetadata::getDataFile)
-                                                             .anyMatch(df -> worldMapTokenService.isWorldMapTokenAuthorizedForDataFileDownload(apiToken, df));
+        if (!GuestUser.get().equals(apiTokenUser) &&
+                permissionService.requestOn(createDataverseRequest(apiTokenUser), file).has(Permission.DownloadFile)) {
 
-            if (!isWorldMapTokenPresentAndAuthorized) {
-                return false;
-            }
+            logger.log(Level.FINE, "Token-based auth: user {0} has access rights to file with id: {1}.",
+                    new Object[] { apiTokenUser.getIdentifier(), file.getId() });
+            return true;
+        } else if (!GuestUser.get().equals(sessionUser) &&
+                permissionService.requestOn(createDataverseRequest(sessionUser), file).has(Permission.DownloadFile)) {
 
-            logger.fine("WorldMap token-based auth: Token is valid for the requested datafile");
+            logger.log(Level.FINE, "Session-based auth: user {0} has access rights to file with id: {1}.",
+                    new Object[] { sessionUser.getIdentifier(), file.getId() });
             return true;
+        } else if (permissionService.requestOn(createDataverseRequest(GuestUser.get()), file).has(Permission.DownloadFile)) {
 
+            logger.log(Level.FINE, "Guest user has access rights to file with id: {1}.", file.getId());
+            return true;
         }
 
         return false;
     }
 
+    private User getApiTokenUserWithGuestFallbackOnInvalidToken() {
+        return Try.of(this::findUserOrDie)
+                .onFailure(throwable -> logger.log(Level.FINE, "Failed finding user with apiToken", throwable))
+                .getOrElse(GuestUser.get());
+    }
+
+    private User getSessionUserWithGuestFallback() {
```

Comment:
Since method above uses functional-like style maybe consider it here as well with `Optional.ofNullable` or something from vavr?